lib/engine/game/g_1824.rb

```diff
@@ -228,41 +237,126 @@ def init_round
         ])
       end
 
-      def purchasable_companies(_entity = nil)
-        []
+      def purchasable_unsold_companies
+        @companies.reject(&:owner).reject(&:closed?)
+      end
+
+      def regional_bk
+        @bk ||= corporation_by_id('BK')
+      end
+
+      def regional_ms
+        @ms ||= corporation_by_id('MS')
+      end
+
+      def regional_cl
+        @cl ||= corporation_by_id('CL')
+      end
+
+      def regional_sb
+        @sb ||= corporation_by_id('SB')
+      end
+
+      def state_sd
+        @sd ||= corporation_by_id('SD')
+      end
+
+      def state_ug
+        @ug ||= corporation_by_id('UG')
+      end
+
+      def state_kk
+        @kk ||= corporation_by_id('KK')
       end
 
       def setup
-        g_trains = @depot.upcoming.select { |t| g_train?(t) }
-        coal_railways.each do |coalcorp|
-          train = g_trains.shift
-          buy_train(coalcorp, train, :free)
-          coalcorp.spend(120, @bank, check_cash: false)
-        end
+        @two_train_bought = false
 
         @minors.each do |minor|
           hex = hex_by_id(minor.coordinates)
           hex.tile.cities[minor.city].place_token(minor, minor.next_token)
         end
+
+        # Reserve the last share (which is the 20% presidency) to
+        # have it as exchange for associated coal railway
+        associated_regional_railways.each do |regional|
+          regional.shares.find(&:president).buyable = false
+          regional.floatable = false
+        end
       end
 
       def ipo_name(entity)
-        return 'Treasury' if coal_railways.include?(entity)
+        return 'Treasury' if coal_railway?(entity)
 
         'IPO'
       end
 
       def can_par?(corporation, parrer)
-        super && !corporation.all_abilities.find { |a| a.type == :no_buy }
+        super && !corporation.all_abilities.find { |a| a.type == :no_buy } &&
```

Comment:
all_ability.any? or none instead of find beacuse you don't use the return type

Reply:
I changed it to none? 
Also realized there was a bug with the regional so I refactored that statement, and a bit of the code as well.